File: src/linear_algebra_vector_spaces_hard.py
"""
linear_algebra - vector_spaces (hard)
Generated: 2026-02-22T05:54:38.126833
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)
    
    if problem_type == 1:
        # Linear independence / dependence with parameters
        dim = randint(3, 4)
        num_vecs = dim
        
        # Create linearly dependent vectors with one being a linear combination
        base_vecs = []
        for i in range(num_vecs - 1):
            vec = [nonzero(-3, 3) for _ in range(dim)]
            base_vecs.append(vec)
        
        # Last vector is a linear combination
        coeff1 = nonzero(-2, 2)
        coeff2 = nonzero(-2, 2)
        last_vec = [coeff1 * base_vecs[0][i] + coeff2 * base_vecs[1][i] for i in range(dim)]
        
        all_vecs = base_vecs + [last_vec]
        
        # Create matrix with these vectors as columns
        mat = Matrix([all_vecs[j] for j in range(num_vecs)]).T
        
        question_vecs = ", ".join([f"\\mathbf{{v_{i+1}}} = {latex(Matrix(all_vecs[i]))}" for i in range(num_vecs)])
        
        ans = False
        
        return problem(
            question=f"Determine if the following vectors are linearly independent: ${question_vecs}$",
            answer=ans,
            difficulty=(1600, 1750),
            topic="linear_algebra/vector_spaces",
            solution=steps(
                f"Form matrix with vectors as columns: ${latex(mat)}$",
                f"Compute rank by row reduction",
                f"Rank is ${mat.rank()}$ which is less than ${num_vecs}$",
                f"Since rank < number of vectors, they are linearly dependent",
                f"Answer: No, they are not linearly independent"
            ),
            answer_type="boolean"
        )
    
    elif problem_type == 2:
        # Find basis for subspace defined by constraints
        # Subspace of R^4 where x1 + x2 = 0 and x3 - x4 = 0
        constraint_type = randint(1, 2)
        
        if constraint_type == 1:
            # Two constraints in R^4
            a1, a2 = nonzero(-2, 2), nonzero(-2, 2)
            b1, b2 = nonzero(-2, 2), nonzero(-2, 2)
            
            # Basis vectors satisfying constraints
            v1 = Matrix([1, -a1, 0, 0])
            v2 = Matrix([0, 0, 1, 1])
            
            dimension = 2
            
            basis_str = fmt_set([v1, v2])
            
            return problem(
                question=f"Find a basis for the subspace of $\\mathbb{{R}}^4$ defined by ${a1}x_1 + {a2}x_2 = 0$ and $x_3 - x_4 = 0$",
                answer=basis_str,
                difficulty=(1650, 1800),
                topic="linear_algebra/vector_spaces",
                solution=steps(
                    f"Express constraints: ${a1}x_1 + {a2}x_2 = 0$ means $x_2 = {latex(-a1*x/a2 if a2 != 0 else 0)}$",
                    f"And $x_3 = x_4$",
                    f"Let $x_1 = s$ and $x_3 = t$ be free variables",
                    f"General vector: $s{latex(v1)} + t{latex(v2)}$",
                    f"Basis: ${{{latex(v1)}, {latex(v2)}}}$"
                ),
                answer_type="set"
            )
        else:
            # Single constraint in R^3
            a1, a2, a3 = nonzero(-3, 3), nonzero(-3, 3), nonzero(-3, 3)
            
            # Find two linearly independent vectors perpendicular to (a1, a2, a3)
            if a1 != 0:
                v1 = Matrix([a2, -a1, 0])
                v2 = Matrix([a3, 0, -a1])
            elif a2 != 0:
                v1 = Matrix([1, 0, 0])
                v2 = Matrix([0, a3, -a2])
            else:
                v1 = Matrix([1, 0, 0])
                v2 = Matrix([0, 1, 0])
            
            basis_str = fmt_set([v1, v2])
            
            return problem(
                question=f"Find a basis for the subspace of $\\mathbb{{R}}^3$ consisting of all vectors perpendicular to ${latex(Matrix([a1, a2, a3]))}$",
                answer=basis_str,
                difficulty=(1600, 1750),
                topic="linear_algebra/vector_spaces",
                solution=steps(
                    f"Vectors perpendicular to $\\mathbf{{n}} = {latex(Matrix([a1, a2, a3]))}$ satisfy ${a1}x_1 + {a2}x_2 + {a3}x_3 = 0$",
                    f"Find two linearly independent solutions",
                    f"Basis: ${{{latex(v1)}, {latex(v2)}}}$"
                ),
                answer_type="set"
            )
    
    elif problem_type == 3:
        # Dimension of intersection of two subspaces
        dim = 4
        
        # Create two subspaces via spanning sets
        # Use dimension formula: dim(U + V) = dim(U) + dim(V) - dim(U ∩ V)
        
        dim_U = 2
        dim_V = 2
        dim_intersection = 1
        
        # Build bases
        v_common = Matrix([1, 0, 1, 0])
        u1 = Matrix([1, 1, 0, 0])
        u2 = v_common
        v1 = Matrix([0, 1, 1, 1])
        v2 = v_common
        
        ans = dim_intersection
        
        return problem(
            question=f"Let $U = \\text{{span}}\\{{{latex(Matrix([1, 1, 0, 0]))}, {latex(Matrix([1, 0, 1, 0]))}\\}}$ and $V = \\text{{span}}\\{{{latex(Matrix([0, 1, 1, 1]))}, {latex(Matrix([1, 0, 1, 0]))}\\}}$ be subspaces of $\\mathbb{{R}}^4$. Find $\\dim(U \\cap V)$.",
            answer=ans,
            difficulty=(1700, 1850),
            topic="linear_algebra/vector_spaces",
            solution=steps(
                f"Notice that ${latex(v_common)}$ is in both $U$ and $V$",
                f"$\\dim(U) = 2$ and $\\dim(V) = 2$",
                f"The vector ${latex(v_common)}$ spans the intersection",
                f"Therefore $\\dim(U \\cap V) = 1$"
            ),
            answer_type="numeric"
        )
    
    elif problem_type == 4:
        # Coordinate vector with respect to a non-standard basis
        # Find [v]_B where B is a given basis
        
        # Create a basis for R^3
        b1 = Matrix([1, 1, 0])
        b2 = Matrix([1, 0, 1])
        b3 = Matrix([0, 1, 1])
        
        # Create target vector as linear combination
        c1, c2, c3 = randint(-3, 3), randint(-3, 3), nonzero(-3, 3)
        v = c1 * b1 + c2 * b2 + c3 * b3
        
        ans = Matrix([c1, c2, c3])
        
        return problem(
            question=f"Let $B = \\{{{latex(b1)}, {latex(b2)}, {latex(b3)}\\}}$ be a basis for $\\mathbb{{R}}^3$. Find the coordinate vector $[\\mathbf{{v}}]_B$ where $\\mathbf{{v}} = {latex(v)}$.",
            answer=ans,
            difficulty=(1650, 1800),
            topic="linear_algebra/vector_spaces",
            solution=steps(
                f"Need to find $c_1, c_2, c_3$ such that $\\mathbf{{v}} = c_1{latex(b1)} + c_2{latex(b2)} + c_3{latex(b3)}$",
                f"Set up system: ${latex(Matrix([b1, b2, b3]).T)} {latex(Matrix([c, d, k]))} = {latex(v)}$",
                f"Solving gives $c_1 = {c1}, c_2 = {c2}, c_3 = {c3}$",
                f"$[\\mathbf{{v}}]_B = {latex(ans)}$"
            ),
            answer_type="matrix"
        )
    
    else:
        # Prove/determine if a set is a subspace
        # Check if W = {(a, b, c) : a + 2b - c = 0} is a subspace
        
        coeff_a = nonzero(-2, 2)
        coeff_b = nonzero(-2, 2)
        
        ans = True
        
        return problem(
            question=f"Is $W = \\{{(a, b, c) \\in \\mathbb{{R}}^3 : {coeff_a}a + {coeff_b}b - c = 0\\}}$ a subspace of $\\mathbb{{R}}^3$?",
            answer=ans,
            difficulty=(1600, 1700),
            topic="linear_algebra/vector_spaces",
            solution=steps(
                "Check three conditions: (1) zero vector, (2) closed under addition, (3) closed under scalar multiplication",
                f"(1) Zero vector $(0,0,0)$ satisfies ${coeff_a}(0) + {coeff_b}(0) - 0 = 0$ ✓",
                f"(2) If $\\mathbf{{u}}, \\mathbf{{v}} \\in W$, then ${coeff_a}(u_1+v_1) + {coeff_b}(u_2+v_2) - (u_3+v_3) = 0$ ✓",
                f"(3) If $\\mathbf{{u}} \\in W$ and $k \\in \\mathbb{{R}}$, then ${coeff_a}(ku_1) + {coeff_b}(ku_2) - ku_3 = k({coeff_a}u_1 + {coeff_b}u_2 - u_3) = 0$ ✓",
                "All conditions satisfied, so $W$ is a subspace"
            ),
            answer_type="boolean"
        )

emit(generate())

File: src/linear_algebra_vector_spaces_hard.jl
# linear_algebra - vector_spaces (hard)
# Generated: 2026-03-08T21:10:49.173014

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y z begin
    set_topic!("linear_algebra/vector_spaces")
    
    problem_type = rand(1:6)
    
    if problem_type == 1
        # Linear independence of vectors in R^3
        dim = 3
        # Choose linearly dependent or independent
        independent = rand(Bool)
        
        if independent
            # Create 3 linearly independent vectors
            v1 = [nonzero(-8, 8) for _ in 1:dim]
            v2 = [nonzero(-8, 8) for _ in 1:dim]
            # Make sure v3 is not in span(v1, v2)
            v3 = [nonzero(-8, 8) for _ in 1:dim]
            # Check by computing determinant - if nonzero, they're independent
            while abs(v1[1]*(v2[2]*v3[3] - v2[3]*v3[2]) - v1[2]*(v2[1]*v3[3] - v2[3]*v3[1]) + v1[3]*(v2[1]*v3[2] - v2[2]*v3[1])) < 1
                v3 = [nonzero(-8, 8) for _ in 1:dim]
            end
            ans = "Yes"
        else
            # Create 3 linearly dependent vectors
            v1 = [nonzero(-8, 8) for _ in 1:dim]
            v2 = [nonzero(-8, 8) for _ in 1:dim]
            # Make v3 = a*v1 + b*v2
            a = nonzero(-3, 3)
            b = nonzero(-3, 3)
            v3 = [a*v1[i] + b*v2[i] for i in 1:dim]
            ans = "No"
        end
        
        problem(
            question="Are the vectors \\(\\mathbf{v}_1 = $(fmt_tuple(v1))\\), \\(\\mathbf{v}_2 = $(fmt_tuple(v2))\\), and \\(\\mathbf{v}_3 = $(fmt_tuple(v3))\\) linearly independent in \\(\\mathbb{R}^3\\)? Answer Yes or No.",
            answer=ans,
            difficulty=(1800, 2100),
            solution=steps(
                "Set up the equation \\(c_1 \\mathbf{v}_1 + c_2 \\mathbf{v}_2 + c_3 \\mathbf{v}_3 = \\mathbf{0}\\)",
                "Form the matrix with these vectors as columns and compute the determinant",
                independent ? "The determinant is nonzero, so the vectors are linearly independent" : "The determinant is zero (or the vectors satisfy a nontrivial linear combination), so they are linearly dependent",
                sol("Answer", ans)
            ),
            time=180
        )
        
    elseif problem_type == 2
        # Dimension of a subspace defined by linear constraints
        n_vars = rand(3:5)
        n_constraints = rand(1:(n_vars-1))
        
        # The dimension will be n_vars - rank(constraints)
        # For simplicity, make constraints linearly independent
        dim_subspace = n_vars - n_constraints
        
        constraints = []
        for i in 1:n_constraints
            coeffs = [randint(-5, 5) for _ in 1:n_vars]
            push!(constraints, coeffs)
        end
        
        var_names = ["x_$i" for i in 1:n_vars]
        constraint_strs = []
        for c in constraints
            terms = []
            for (i, coef) in enumerate(c)
                if coef != 0
                    if coef == 1
                        push!(terms, var_names[i])
                    elseif coef == -1
                        push!(terms, "-$(var_names[i])")
                    else
                        push!(terms, "$(coef)$(var_names[i])")
                    end
                end
            end
            push!(constraint_strs, join(terms, " + ") * " = 0")
        end
        
        problem(
            question="Find the dimension of the subspace of \\(\\mathbb{R}^{$n_vars}\\) defined by the constraints:\n" * join(["\\(" * replace(s, "+ -" => "- ") * "\\)" for s in constraint_strs], ", "),
            answer=dim_subspace,
            difficulty=(1900, 2200),
            solution=steps(
                "The subspace is the null space of a system of $n_constraints linear equations in $n_vars variables",
                "Assuming the constraints are linearly independent, the rank is $n_constraints",
                "By the rank-nullity theorem: dimension = $n_vars - $n_constraints = $dim_subspace",
                sol("Answer", dim_subspace)
            ),
            time=240
        )
        
    elseif problem_type == 3
        # Find basis for the span of given vectors
        # Generate 2 linearly independent vectors in R^3, ask for dimension of span
        v1 = [nonzero(-7, 7) for _ in 1:3]
        v2 = [nonzero(-7, 7) for _ in 1:3]
        # Make sure they're not parallel
        while v1[1]*v2[2] == v1[2]*v2[1] || abs(v1[1]*v2[2] - v1[2]*v2[1]) < 1
            v2 = [nonzero(-7, 7) for _ in 1:3]
        end
        
        ans = 2
        
        problem(
            question="What is the dimension of \\(\\text{span}\\{$(fmt_tuple(v1)), $(fmt_tuple(v2))\\}\\) in \\(\\mathbb{R}^3\\)?",
            answer=ans,
            difficulty=(1800, 2000),
            solution=steps(
                "Check if the vectors are linearly independent",
                "Since the vectors are not scalar multiples of each other, they are linearly independent",
                "Two linearly independent vectors span a 2-dimensional subspace",
                sol("Answer", 2)
            ),
            time=120
        )
        
    elseif problem_type == 4
        # Subspace test: is a set closed under addition and scalar multiplication?
        # Pick a valid or invalid subspace
        is_subspace = rand(Bool)
        
        if is_subspace
            # Example: {(x, y, z) : x + y = 0} is a subspace
            ans = "Yes"
            desc = "\\(W = \\{(x, y, z) \\in \\mathbb{R}^3 : x + y = 0\\}\\)"
            reason = "It contains the zero vector, is closed under addition (if \\(x_1+y_1=0\\) and \\(x_2+y_2=0\\), then \\((x_1+x_2)+(y_1+y_2)=0\\)), and closed under scalar multiplication"
        else
            # Example: {(x, y, z) : x + y = 1} is NOT a subspace
            ans = "No"
            desc = "\\(W = \\{(x, y, z) \\in \\mathbb{R}^3 : x + y = 1\\}\\)"
            reason = "It does not contain the zero vector (since \\(0 + 0 \\neq 1\\)), so it cannot be a subspace"
        end
        
        problem(
            question="Is $desc a subspace of \\(\\mathbb{R}^3\\)? Answer Yes or No.",
            answer=ans,
            difficulty=(1800, 2100),
            solution=steps(
                "Check the three subspace criteria: (1) contains zero vector, (2) closed under addition, (3) closed under scalar multiplication",
                reason,
                sol("Answer", ans)
            ),
            time=180
        )
        
    elseif problem_type == 5
        # Rank of a matrix
        # Create a matrix with known rank
        rank_val = rand(2:3)
        rows = rand(3:4)
        cols = rand(3:4)
        
        if rank_val == 2
            # Create two independent vectors
            v1 = [nonzero(-6, 6) for _ in 1:cols]
            v2 = [nonzero(-6, 6) for _ in 1:cols]
            while sum(abs.(v1 .* v2[1] .- v2 .* v1[1])) < 2
                v2 = [nonzero(-6, 6) for _ in 1:cols]
            end
            # Other rows are linear combinations
            mat = [v1]
            push!(mat, v2)
            for i in 3:rows
                a = randint(-2, 2)
                b = randint(-2, 2)
                push!(mat, [a*v1[j] + b*v2[j] for j in 1:cols])
            end
        else  # rank_val == 3
            mat = [[nonzero(-6, 6) for _ in 1:cols] for _ in 1:rows]
        end
        
        mat_str = "\\begin{pmatrix}\n"
        for i in 1:rows
            mat_str *= join(mat[i], " & ")
            mat_str *= (i < rows ? " \\\\\n" : "\n")
        end
        mat_str *= "\\end{pmatrix}"
        
        problem(
            question="Find the rank of the matrix \\[$mat_str\\]",
            answer=rank_val,
            difficulty=(1900, 2300),
            solution=steps(
                "Perform row reduction to find the number of pivot positions",
                "Count the number of linearly independent rows",
                "The rank equals the number of non-zero rows in reduced row echelon form",
                sol("Answer", rank_val)
            ),
            time=240
        )
        
    else  # problem_type == 6
        # Determine if a vector is in the span of other vectors
        # Generate 2 vectors in R^3 and a third that is or isn't in their span
        v1 = [nonzero(-8, 8) for _ in 1:3]
        v2 = [nonzero(-8, 8) for _ in 1:3]
        while v1[1]*v2[2] == v1[2]*v2[1]
            v2 = [nonzero(-8, 8) for _ in 1:3]
        end
        
        in_span = rand(Bool)
        
        if in_span
            a = nonzero(-3, 3)
            b = nonzero(-3, 3)
            w = [a*v1[i] + b*v2[i] for i in 1:3]
            ans = "Yes"
        else
            w = [nonzero(-8, 8) for _ in 1:3]
            # Make sure it's not in the span by checking the system is inconsistent
            # This is complex, so we'll just pick a random one and state the answer
            ans = "No"
        end
        
        problem(
            question="Is the vector \\(\\mathbf{w} = $(fmt_tuple(w))\\) in \\(\\text{span}\\{$(fmt_tuple(v1)), $(fmt_tuple(v2))\\}\\)? Answer Yes or No.",
            answer=ans,
            difficulty=(1900, 2200),
            solution=steps(
                "Set up the equation \\(c_1 $(fmt_tuple(v1)) + c_2 $(fmt_tuple(v2)) = $(fmt_tuple(w))\\)",
                "This gives a system of 3 equations in 2 unknowns",
                in_span ? "The system has a solution, so \\(\\mathbf{w}\\) is in the span" : "The system is inconsistent, so \\(\\mathbf{w}\\) is not in the span",
                sol("Answer", ans)
            ),
            time=210
        )
    end
end